### PR TITLE
Initial Implementation of XML Format

### LINKED ISSUE #446: XML Format Support
This is a placeholder to add support for the XML Format that is currently in the process of being formalized in the CE working group.

I am in the process of building an initial implementation as the format specification evolves as a way to validate the specification

### PR DESCRIPTION
**Pull request by JemDay**
- I believe this is "spec compliant - I was wring this at the same time I was writing the spec.
- We *may* want to consider a streaming parsing implementation at some point, but I'm not sure that's needed now.
- Contains +ve and -ve tests along with some round-trip verifications.
- Implementation is contained and not visible outside the package/module.
- Comments welcome :-)

Closes #446

### CODE REVIEW

File: docs/xml.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+// Construct the event
+CloudEvent event = CloudEventBuilder.v1()
+    .withId("hello")
+    .withType("example.xml")
+    .withSource(URI.create("http://localhost"))
+    .withData(myData)
+    .build();
+
+
```
[the remark this answers is not in the document]
Comment (JemDay) [reply docs/xml.md:74]: ack

File: formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+                final String attrType = extractAttributeType(e);
+                final String attrValue = e.getTextContent();
+
+                // Check if this is a Required or Optional attribute
+                if (specVersion.getAllAttributes().contains(attrName)) {
+                    // Yep .. Just write it out.
+                    writer.withContextAttribute(attrName, attrValue);
+                } else {
+                    if (XMLConstants.XML_DATA_ELEMENT.equals(attrName)) {
+                        // Just remember the data node for now..
```
[the remark this answers is not in the document]
Comment (JemDay) [reply formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java:88]: Ack

File: docs/xml.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import io.cloudevents.core.builder.CloudEventBuilder;
+import io.cloudevents.xml.XMLFormat;
+
+CloudEvent event = CloudEventBuilder.v1()
+    .withId("hello")
+    .withType("example.xml")
+    .withSource(URI.create("http://localhost"))
+    .build();
+
+byte[]serialized = EventFormatProvider
```
[the remark this answers is not in the document]
Comment (JemDay) [reply docs/xml.md:41]: Ack

File: formats/xml/pom.xml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            <classifier>tests</classifier>
+            <type>test-jar</type>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.xmlunit</groupId>
+            <artifactId>xmlunit-core</artifactId>
+            <version>2.9.0</version>
```
[the remark this answers is not in the document]
Comment (JemDay) [reply formats/xml/pom.xml:69]: ack

File: formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+
+    // Private Methods --------------------------------------------------------
+
+    /**
+     * Get the first child {@link Element} of an {@link Element}
+     *
+     * @param e
+     * @return The first child, or NULL if there isn't one.
+     */
+    private Element findFirstElement(Element e) {
```
[the remark this answers is not in the document]
Comment (JemDay) [reply formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java:138]: ack

File: docs/xml.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+The `EventFormatProvider` will resolve automatically the `XMLFormat` using the
+`ServiceLoader` APIs.
+
+XML Document data handling is supported via the `XMLCloudEventData`
+facility. This convenience wrapper can be used with `any` other supported
+format.
+
+```java
+import org.w3c.dom.Document;
```
[the remark this answers is not in the document]
Comment (JemDay) [reply docs/xml.md:55]: That's always been a part of the JDK.

Out of transparency I've also though about creating another variant of the JSON Format that uses the native JDK facilities with no additional dependencies.

File: formats/xml/src/main/resources/META-INF/services/io.cloudevents.core.format.EventFormat [hunk 7]
```diff
@@ -0,0 +1 @@
+io.cloudevents.protobuf.XMLFormat
```
[the remark this answers is not in the document]
Comment (JemDay) [reply formats/xml/src/main/resources/META-INF/services/io.cloudevents.core.format.EventFormat:1]: Ack .. nice catch ...

File: kafka/pom.xml [hunk 8]
```diff
@@ -23,15 +23,15 @@
...
+        <version>2.5.0-SNAPSHOT</version>
     </parent>
 
     <artifactId>cloudevents-kafka</artifactId>
     <name>CloudEvents - Kafka Binding</name>
     <packaging>jar</packaging>
 
     <properties>
-        <kafka.version>2.5.0</kafka.version>
+        <kafka.version>3.0.0</kafka.version>
```
Comment (pierDipi) [inline kafka/pom.xml:34]: There are some unrelated changes like this one, would you mind rebasing?

Comment (JemDay) [reply kafka/pom.xml]: @pierDipi - That's for your patience - rebased & pushed

File: formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+                throw CloudEventRWException.newInvalidDataType(el.getLocalName(), "Unexpected child element(s)");
+            }
+        }
+
+        // Finally, ensure we only see each CE Attribute once...
+        try {
+            ceAttributeTracker.trackOccurrence(localName);
+        } catch (IllegalStateException e){
+            throw CloudEventRWException.newOther(e);
+        }
```
Comment (pierDipi) [inline formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java:294]: Would a Map-like API work here, instead of an exception-based control flow?
```java
if (ceAttributeTracker.trackOccurrence(localName)) {
   throw CloudEventRWException.new...
}
```

File: core/src/main/java/io/cloudevents/core/impl/BaseCloudEventBuilder.java [hunk 10]
```diff
@@ -124,6 +124,9 @@ public SELF withExtension(@Nonnull String key, @Nonnull String value) {
         return self;
     }
 
+    // @TODO - I think this method should be removed/deprecated
+    // **Number** Is NOT a valid CE Context atrribute type.
+
```
[the remark this answers is not in the document]
Comment (JemDay) [reply core/src/main/java/io/cloudevents/core/impl/BaseCloudEventBuilder.java:129]: Agree ... we should probably flag it as deprecated to steer people away from using it .. but that's a separate issue I think.

File: formats/xml/src/main/java/io/cloudevents/xml/XMLSerializer.java [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        private boolean isXMLContent(String contentType) {
+            if (contentType == null) {
+                return false;
+            }
+
+            return "application/xml".equals(contentType)
+                || "text/xml".equals(contentType)
+                || contentType.endsWith("+xml")
+                ;
+        }
```
[the remark this answers is not in the document]
Comment (JemDay) [reply formats/xml/src/main/java/io/cloudevents/xml/XMLSerializer.java:154]: I'll change to make it consistent ... however, those type of charset statements only really make sense with a binary payload ... the content encoding of the xml document has to be consistent ..

File: formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java (hunk 9 above)
Comment (JemDay) [reply formats/xml/src/main/java/io/cloudevents/xml/XMLDeserializer.java:294]: I will make this change ...

File: formats/xml/src/main/java/io/cloudevents/xml/XMLUtils.java [hunk 12]
```diff
@@ -36,9 +36,15 @@
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 class XMLUtils {
 
+    private static final Pattern XML_PATTERN = Pattern.compile("^(application|text)\\/([a-zA-Z]+\\+)?xml(;.*)*$");
+    private static final Pattern TEXT_PATTERN = Pattern.compile("^application\\/([a-zA-Z]+\\+)?(xml|json)(;.*)*$");
```
Comment (pierDipi) [inline formats/xml/src/main/java/io/cloudevents/xml/XMLUtils.java:45]: We're adding `json` too because it's a common / well known "text-based format", right?

Comment (JemDay) [reply formats/xml/src/main/java/io/cloudevents/xml/XMLUtils.java:45]: Yes ... any `text/*` content-types are caught in the utility function without using the regex (to keep it simpler!).

**Review by pierDipi — APPROVED** [review]
Thanks!

LGTM
